### ISSUE TITLE
GPU support wanted / relevant?

### ISSUE THREAD

**Issue by sesam** [cathugger/mkp224o#13]
I see no mention of GPU support, if it's relevant, PRs wanted, etc.

Here's another ed25519 vanity generator that might be interesting to compare with: https://docs.rs/crate/nano-vanity/0.2.1

**Comment by cathugger** [comment]
There was https://github.com/cathugger/mkp224o/issues/4 but basically I don't really have hardware to test with and much experience with programming GPUs :\
And I'm not really sure how it'd integrate with stuff like filters, and is this project right scope for that.
Thanks for mentioning https://github.com/PlasmaPower/nano-vanity looks like it's doing pretty much same kind of thing.
Should we filter inside GPU? Or send all generated keys to CPU for that? Or mixed approach with subset of wider filters in GPU and full filtering in CPU side?
How should it relate to main mkp224o application? Be part of it, or separate executable? (probably separate, different kind of dependencies)
mkp224o is CC0 (with some parts put into Public Domain), but since https://github.com/PlasmaPower/nano-vanity is BSD, and I don't think it's worth redoing it, that part of codebase would be BSD licensed.
Should it be in this project scope, or separate repository should be made? Not sure..

Lot of uncertainty from my side... I'll need to think about it, but I myself currently not doing any code for this, so incase anyone wants to work on it, they shouldn't hesitate.

**Comment by Cyclic3** [comment]
Have you tried [pocl](http://portablecl.org/)? That works with just a CPU, so you should be able to test with it.

**Comment by Baaleos** [comment]
In regards to the question of when to do the filtering - Scallion - a .net GPU generator for onion addresses I think just generates a work load, processes the GPG keys and then filters on the cpu. 
The important thing is to keep the GPU hard at work, if it runs out of work it then throttles down.

**Comment by cathugger** [comment]
I'm too busy with other irl stuff, and have too little interest in doing this myself (lack of actual powerful GPU isn't motivating, because I personally won't benefit from it, and even testing might get unnecessarily tedious (unless I try with pocl but that's one more thing to learn to play with..)).
Will leave open, as there is some good knowledge in this issue in case someone might be interested doing this.

**Comment by exander77** [comment]
Is here a documentation how are the new adresses generated?

**Comment by cathugger** [comment]
currently no but reading code shouldn't be hard.
essentially we generate ed25519 pubkeys as quick as possible filtering them as quickly as possible and once we get match we reseed private key, and format pubkey x privkey pair in format tor accepts and dump that to filesystem. or output in yaml-ish format.
there are few ways to calculate public keys as quick as possible.
there are also few ways to filter and it depends on amount of filters and their properties which way to do it works best. which is why it's configurable at compile time.

**Comment by whalehub** [comment]
@cathugger Would you be open to working in this if someone sponsored you an RTX GPU?

**Comment by whalehub** [comment]
@cathugger Would you be open to working on this if someone sponsored you an RTX GPU?

**Comment by cathugger** [comment]
@whalehub It'd increase likehood. But then I don't even have proper motherboard for that (my only serious working machine right now is laptop). Might get some better hardware this summer, maybe.
Why is nobody which isn't me doing it, tho, I mean there definitely must be people capable and having all the right hardware, and it shouldn't even be too hard. There's even https://github.com/PlasmaPower/nano-vanity/blob/master/src/opencl/curve25519.cl, also there are useful tips in https://eprint.iacr.org/2014/198.pdf.

**Comment by Scripter17** [comment]
I don't mean to be rude, but has anyone made any progress on this front? I can't find any other software that does what the issue proposes.  
It's been a bit over a year and a half since this issue started, and just a few days over a year since the last comment. I highly doubt that *no one* has done anything with this in that time.

**Comment by cathugger** [comment]
as I've indicated earlier I'm not working on it atm.
it kinda looks a bit odd that no one's doing it too.

**Comment by dr-bonez** [comment]
I have built one that works with cuda in Rust.
https://github.com/dr-bonez/tor-v3-vanity

**Comment by GIJack** [comment]
I could definitely use some OpenCL support. Especially multi-device...

I have two GPUs and would make my life easier...

**Comment by Dibbyo456** [comment]
@dr-bonez which is would be better? AMD 5900x CPU or RTX 3080 GPU?

**Comment by dr-bonez** [comment]
There's definitely something wrong with my generator that is slowing it down, that I haven't had time to debug. For now, I'd recommend continuing to use the cpu version.

**Comment by Dibbyo456** [comment]
Alright. Thank you. 🙏

**Comment by mooleshacat** [comment]
> Why is nobody who isn't me doing it, tho, I mean there definitely must be people capable and having all the right hardware, and it shouldn't even be too hard.

I have the hardware, just no clue how and I suck at programming. Very limited in programming ability. But I'd say 5950x w/ RTX 3070 is more than enough.

If anyone does code this and need testing let me know :) I'd be glad to test (and use when done)

**Comment by adapt-L** [comment]
I think I still have some code saved from last year when I attempted this with OpenCL. It was very inefficient, orders of magnitude slower than mkp244o (perhaps because it was a copy of supercop or donna that used branching)? This was just a test on intel graphics though, I have not been able to get OpenCL working on amd or nvidia.

It only did filtered a single vanity at a time, by converting the vanity into a bit array, and then converting the length into a bit mask.

I can share if anyone is interested. Maybe someone with OpenCL experience could fix it up...

**Comment by cathugger** [comment]
please do share. i kinda do have capable setup now, just not much time. even if not for me, may be useful for someone else.

**Comment by exander77** [comment]
> I think I still have some code saved from last year when I attempted this with OpenCL. It was very inefficient, orders of magnitude slower than mkp244o (perhaps because it was a copy of supercop or donna that used branching)? This was just a test on intel graphics though, I have not been able to get OpenCL working on amd or nvidia.
> 
> I can share if anyone is interested. Maybe someone with OpenCL experience could fix it up...

Yes, if you share it, We can definitely optimize. I hate interfacing with OpenCL, but I can look into optimizing already existing code.

**Comment by adapt-L** [comment]
@exander77 Here is a link: https://github.com/adapt-L/GenEd
It's just a prototype that generates the keypair and filters it. My CPU (i5-1240P) gives me 0.28676 million keypairs/sec and my integrated GPU gives me 0.0614793 million keypairs/sec

**Comment by Zipdox2** [comment]
I wrote this but I'm having trouble getting it to run on my GPU because Mesa OpenCL support is poor. https://gitlab.com/zipdox/onion-v3-vanity-opencl

**Comment by daoudiyounes** [comment]
any news on that ?

**Comment by Zipdox2** [comment]
> any news on that ?

Not really https://gitlab.freedesktop.org/mesa/mesa/-/issues/9926

**Comment by Meow** [comment]
tl;dr: GPU is slower at this problem than CPU

I've ported the ref10 algorithm to CUDA and did my absolute best to make it GPU-friendly (inlining everything, trying to be careful about pointers, using extremely efficient prefix checking algorithms), and I have determined this to largely be pointless because it can't beat what my CPU is outputting.

mkp224o on the CPU (batched mode, amd64-64-24k algorithm) manages 130M keys per second on 9900X, and 125M keys per second on 7950X, using maybe 200W of power.

My CUDA implementation (ran with kernel config <<<512, 256>>>, each thread doing 1000 key generations and comparisons), which as far as I am aware is leagues faster than what I found on the internet, managed 60M/sec on a RTX4090, using 420W of power. Oh and did I mention that it's using a far less secure randomness algorithm, and doesn't check primarily of keys? Yeah.

According to the NCU profiler, there's potential for maybe a 2x speedup if I bend over backwards and break every bone in my body writing the least readable code of all time, but honestly, not worth it. It'll barely be as fast as my CPU, at 2x more power consumption.

This is a stupidly parallelizable problem, and it seems from the first glance like a GPU is perfect for it, but ed25519 is just a bad algorithm to run on the GPU.

If you really need some stupidly long prefix like 10 letters long (anything 9 letters and below can be generated on regular consumer hardware in less than a week, so just do that lol), just rent a server from hetzner for 200 euros, which would have 48 zen 4 cores, and run this algorithm there. Cheaper than getting a huge CPU yourself. If you need 11 letter prefix or longer, may gods have mercy upon you, because the amount of compute required for that is absolutely unfunny.

**Comment by levinster82** [comment]
@Meow I think you will be surprised at what your 4090 can do with this. https://github.com/cathugger/mkp224o/pull/121 Im testing with a 3070 and getting 850M Keys/sec

**Comment by adapt-L** [comment]
hey @Meow did your tests uncover any pain points for the algo on the GPU? When I gave up, I figured that some sort of hybrid GPU/CPU approach might work, where you break the problem down into chunks like hash generation, etc. and then benchmark on the CPU and GPU to figure out a pipeline that is most performant. Obviously this would be pretty limited by GPU<->CPU bandwidth.

I agree that this should be a stupidly paralellizable task, I think that is why I also wasted some time attempting it.

I also have a powerful epyc server and a real GPU to test on now, so I may take another crack at this. Albeit it's an AMDGPU and still cant get rocm working after all these years. It might be worth it to just try with a WebGPU/Vulkan backend for portability's sake. I have been meaning to learn WebGPU compute shaders anyways. But based on your reply it may not be worth it anyways

It's probably just the case that this task is naturally CPU-bound due to the features the GPU manufacturers tend to optimize.

**Comment by Meow** [comment]
@adapt-L it has been a while but I remember randomness was painful because GPUs suck at secure randomness. And also checking primality of keys but that can be done on the CPU once the GPU crunches a match.
